lib/services/monitoring.rb

```diff
@@ -118,5 +88,58 @@ def self.important_tag?(tag)
       tag = tag.downcase
       !(tag.include?("preview") || tag.include?("rc") || tag.include?("beta"))
     end
+
+    def self.read_changelog(repo_path, client, latest_tag_index)
+      begin
+        changelog_file = client.contents(repo_path)
+                               .filter { |file| file.name.match(Regexp.new("changelog.md", Regexp::IGNORECASE)) }
+                               .first
+      rescue Octokit::Error => e
+        log_error e
+        return nil
+      end
+
+      if changelog_file.nil?
+        return nil
+      end
+
+      changelog_download_url = changelog_file.download_url
+      download = URI.open(changelog_download_url)
+      if download.nil?
+        return nil
+      end
+
+      changelog = download.read
+      unless changelog.nil?
+        changelog_tag_index = changelog.index(/[\s#]+(#{latest_tag}|#{latest_tag.delete_prefix("v")})/)
+        prev_tag = latest_tag_index < tags.size - 1 ? tags[latest_tag_index + 1].name : nil
+        changelod_end_index = prev_tag.nil? ?
+                                changelog.length :
+                                changelog.index(/[\s#]+(#{prev_tag}|#{prev_tag.delete_prefix("v")})/)
+        return changelog[changelog_tag_index..changelod_end_index]
+      end
+
+      nil
+    end
+
+    def self.find_latest_release(releases)
+      latest_release = releases.first
+      releases.each do |release|
+        latest_release = release
+        latest_tag = release.tag_name
+        if latest_release.nil? ? important_tag?(latest_tag) : important_release?(latest_release)
+          break
+        end
+      end
+
+      latest_release
+    end
+
+    def self.release_over_tag?(latest_tag_name, latest_tag_date, latest_release)
+      # 1. the release is less than approximately 4 months older than the tag      #
+      # 2. the tag and the release are of the same versioning level, e.g. rails 7.0.3 vs 7.0.3.1
+      latest_tag_date - latest_release.published_at > 30 * 4 &&
```

Comment:
Rails provides some time related methods, e.g., `4.months`. I'd suggest trying to use these with a proper date class.